dbms/src/Interpreters/InterpreterCreateQuery.cpp

```diff
@@ -562,6 +578,62 @@ BlockIO InterpreterCreateQuery::createTable(ASTCreateQuery & create)
 	return {};
 }
 
+ASTPtr InterpreterCreateQuery::createQueryWithoutCluster(ASTCreateQuery & create) const
+{
+	ASTPtr cloned = create.clone();
+	ASTCreateQuery & tmp = typeid_cast<ASTCreateQuery &>(*cloned);
+	tmp.cluster.clear();
+	if (tmp.database.empty())
+		tmp.database = context.getCurrentDatabase();
+	return cloned;
+}
+
+void InterpreterCreateQuery::writeToZookeeper(ASTPtr query, const std::vector<Cluster::Address> & addrs)
+{
+	auto zookeeper = context.getZooKeeper();
+	ASTCreateQuery & create = typeid_cast<ASTCreateQuery &>(*query);
+	std::string table_name = create.database + "." + create.table;
+
+	for (const auto & addr : addrs)
+	{
+		const std::string & path =
+			"/clickhouse/task_queue/ddl/" +
+			addr.host_name +
+			"/create/" + table_name;
+
+		// TODO catch exceptions
+		zookeeper->createAncestors(path);
+		zookeeper->create(path, formatASTToString(*query), 0);
+	}
+}
+
+BlockIO InterpreterCreateQuery::createTableOnCluster(ASTCreateQuery & create)
+{
+	ClusterPtr cluster = context.getCluster(create.cluster);
+	Cluster::AddressesWithFailover shards = cluster->getShardsWithFailoverAddresses();
+	ASTPtr query_ptr = createQueryWithoutCluster(create);
+	std::string query = formatASTToString(*query_ptr);
+	std::vector<Cluster::Address> failed;
+
+	for (const auto & shard : shards)
+	{
+		for (const auto & addr : shard)
+		{
+			try
+			{
+				ExecuteQuery(addr, query);
+			}
+			catch (...)
+			{
+				failed.push_back(addr);
```

Comment:
Ok. But we also need to collect information from exceptions and then write log,
... and further we need to return a resultset for distributed create query, containing information, that following addresses was processed Ok, and following will be processed asynchronously.